### PR TITLE
add Variational Auto Encoder

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

fix https://github.com/intel-analytics/BigDL/issues/1416

add KL Loss criteria and Gaussian sampling layer, Gaussian Criterion
add VAE example

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
Add vae example

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    log(sigma) + 0.5 *log(2pi) + 0.5 * (x - mu)^2/sigma^2
+    input[1] = mu
+    input[2] = log(sigma^2)
+    */
+    mean.resizeAs(input[Tensor[T]](1)).copy(input(1))
+    vari.resizeAs(input[Tensor[T]](2)).copy(input(2))
+    expVar.resizeAs(input[Tensor[T]](2)).copy(input(2))
+
+    expVar.exp()
+    vari.mul(ev.fromType(0.5)).add(ev.fromType(0.5 * math.log(2 * math.Pi)))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala]: using a buffer tensor with entry one, and using MKL.apxy ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    vari.add(mean.add(ev.fromType(-1), target).pow(ev.fromType(2))
+      .cdiv(expVar).mul(ev.fromType(0.5)))
+
+    output = vari.sum()
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala]: I think using MKL.dot() will be faster

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala (hunk 1 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala]: here just operate with figure, not tensor.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractCriterion
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class GaussianCriterion[@specialized(Float, Double) T: ClassTag](
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/GaussianCriterion.scala]: Where is this used?

**Comment by zhangxiaoli73** [comment]
jekins pass

**Pull request merged**